salt/utils/versions.py

```diff
@@ -162,6 +163,92 @@ def _formatwarning(message,
         warnings.formatwarning = saved
 
 
+def _get_utcnow_date():
+    '''
+    This function exists because we can't easily patch builtin objects when mocking
+    '''
+    return datetime.datetime.utcnow().date()
+
+
+def warn_until_date(date,
+                    message,
+                    category=DeprecationWarning,
+                    stacklevel=None,
+                    _dont_call_warnings=False):
+    '''
+    Helper function to raise a warning, by default, a ``DeprecationWarning``,
+    until the provided ``date``, after which, a ``RuntimeError`` will
+    be raised to remind the developers to remove the warning because the
+    target date has been reached.
+
+    :param date: A ``datetime.date`` or ``datetime.datetime`` instance.
+    :param message: The warning message to be displayed.
+    :param category: The warning class to be thrown, by default
+                     ``DeprecationWarning``
+    :param stacklevel: There should be no need to set the value of
+                       ``stacklevel``. Salt should be able to do the right thing.
+    :param _dont_call_warnings: This parameter is used just to get the
+                                functionality until the actual error is to be
+                                issued. When we're only after the date
+                                checks to raise a ``RuntimeError``.
+    '''
+    _strptime_fmt = '%Y%m%d'
+    if not isinstance(date, (six.string_types, datetime.date, datetime.datetime)):
+        raise RuntimeError(
+            'The \'date\' argument should be passed as a \'datetime.date()\' or '
+            '\'datetime.datetime()\' objects or as string parserable by '
+            '\'datetime.datetime.strptime()\' with the following format \'{}\'.'.format(
+                _strptime_fmt
+            )
+        )
+    elif isinstance(date, six.text_type):
+        date = datetime.datetime.strptime(date, _strptime_fmt)
+
+    # We're really not interested in the time
+    if isinstance(date, datetime.datetime):
+        date = date.date()
+
+    if stacklevel is None:
+        # Attribute the warning to the calling function, not to warn_until_date()
+        stacklevel = 2
+
+    today = _get_utcnow_date()
+    if today >= date:
+        caller = inspect.getframeinfo(sys._getframe(stacklevel - 1))
+        raise RuntimeError(
+            'The warning triggered on filename \'{filename}\', line number '
+            '{lineno}, is supposed to be shown until {date}. Today is '
+            '{today}. Please remove the warning.'.format(
+                filename=caller.filename,
+                lineno=caller.lineno,
+                date=date,
+                today=today,
+            ),
+        )
+
+    if _dont_call_warnings is False:
+        def _formatwarning(message,
```

Comment:
Does `_formatwarning` need to be a nested function?

Reply:
That was copied from the other `warn_until` function, but I guess I can be reused and factored out.

Reply:
Fixed